Fix CRD validation messages

gautierdelorme opened the pull request:
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Update `github.com/go-openapi/validate` to properly show incorrect values in CRD validation messages.

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubernetes/issues/90128

**Special notes for your reviewer**:

The relevant change I did is to run `hack/pin-dependency.sh github.com/go-openapi/validate c0f6196cdffec6c4d3d3c941927e5483bf607cd4` to only get the relevant fix without any other dependency upgrades. This is a more surgical approach as recommended by @thockin below.

**Does this PR introduce a user-facing change?**:

```release-note
Fixed a bug that prevents CRD validation messages to show incorrect values.
```

gautierdelorme commented:
/sig api-machinery

gautierdelorme commented:
/assign @thockin

nikhita commented:
/ok-to-test

liggitt commented:
looks like `hack/update-vendor.sh` is required

cc @kubernetes/licensing for review once the dependencies CI job passes

gautierdelorme commented:
@liggitt Thanks for taking a look. I ran that command already but it failed in `update-vendor-licenses.sh`:

```console
+++ [0616 14:16:03] vendor: updating vendor/LICENSES
+ shift
+ hack/update-vendor-licenses.sh
++ kube::log::errexit
++ local err=9
++ set +o
++ grep -qe '-o errexit'
++ set +o xtrace
!!! Error in hack/update-vendor.sh:351
  Error in hack/update-vendor.sh:351. 'hack/update-vendor-licenses.sh >> "${LOG_FILE}" 2>&1' exited with status 9
Call stack:
  1: hack/update-vendor.sh:351 main(...)
Exiting with status 1
```

sttts commented on vendor/github.com/asaskevich/govalidator/patterns.go:
```diff
@@ -4,49 +4,51 @@ import "regexp"
-    hasWhitespace     string = ".*[[:space:]]"
-    hasWhitespaceOnly string = "^[[:space:]]+$"
+	Email             string = "^(((([a-zA-Z]|\\d|[!#\\$%&'\\*\\+\\-\\/=\\?\\^_`{\\|}~]|[\\x{00A0}-\\x{D7FF}\\x{F900}-\\x{FDCF}\\x{FDF0}-\\x{FFEF}])+(\\.([a-zA-Z]|\\d|[!#\\$%&'\\*\\+\\-\\/=\\?\\^_`{\\|}~]|[\\x{00A0}-\\x{D7FF}\\x{F900}-\\x{FDCF}\\x{FDF0}-\\x{FFEF}])+)*)|((\\x22)((((\\x20|\\x09)*(\\x0d\\x0a))?(\\x20|\\x09)+)?(([\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x7f]|\\x21|[\\x23-\\x5b]|[\\x5d-\\x7e]|[ ...
+	CreditCard        string = "^(?:4[0-9]{12}(?:[0-9]{3})?|5[1-5][0-9]{14}|(222[1-9]|22[3-9][0-9]|2[3-6][0-9]{2}|27[01][0-9]|2720)[0-9]{12}|6(?:011|5[0-9][0-9])[0-9]{12}|3[47][0-9]{13}|3(?:0[0-5]|[68][0-9])[0-9]{11}|(?:2131|1800|35\\d{3})\\d{11}|6[27][0-9]{14})$"
```
which consequence does this change have on our API? I don't think we can merge a different format like this.

/hold

fedebongio commented:
/assign @sttts

sttts commented:
Discussion continued in Slack https://kubernetes.slack.com/archives/C0EG7JC6T/p1592312250277400?thread_ts=1592287631.273100&cid=C0EG7JC6T

gautierdelorme commented:
/test pull-kubernetes-dependencies

gautierdelorme commented:
/retest

thockin commented:
This drags in yaml.v3 - can we kill off yaml.v2 or do we really want to tolerate both deps?

thockin commented:
Is there any more surgical change we can make - with less dependencies?

gautierdelorme commented:
@thockin I actually paid a lot of attention at this when fixing `go-openapi`. If you look at the change I made in https://github.com/go-openapi/errors/commit/fe99469e5f75d5a6f4850b7e1c039889545b3146 and https://github.com/go-openapi/validate/commit/272b5927f3d2b9b256af6f6c4d688c5fd7530a3d you'll see I didn't bump any dependencies on purpose. Deps got upgraded later by `go-openapi` maintainers before creating the new release including my changes. Do you have suggestion that could help moving this PR forward without having to re-implement code from `govalidator`?

thockin commented:
Can we move everything over to yaml.v3 first and then do this (different commits in same PR or different PRs)?  I don't know what's up with v2 vs v3

neolit123 commented:
/remove-sig cluster-lifecycle

gautierdelorme committed:
fix CRD validation messages

hack/pin-dependency.sh github.com/go-openapi/validate c0f6196cdffec6c4d3d3c941927e5483bf607cd4

Signed-off-by: Gautier Delorme <gautier.delorme@gmail.com>

gautierdelorme committed:
fix tests

Signed-off-by: Gautier Delorme <gautier.delorme@gmail.com>

gautierdelorme commented:
@sttts PTAL - I've refactored the change using a more surgical approach as recommended by @thockin.

sttts commented on go.mod:
```diff
@@ -256,15 +254,15 @@ replace (
 	github.com/go-openapi/strfmt => github.com/go-openapi/strfmt v0.19.3
 	github.com/go-openapi/swag => github.com/go-openapi/swag v0.19.5
-	github.com/go-openapi/validate => github.com/go-openapi/validate v0.19.5
+	github.com/go-openapi/validate => github.com/go-openapi/validate v0.19.9-0.20200613183935-c0f6196cdffe
```
is there no release yet?

sttts replied:
There is 0.19.12 already. Maybe we should use that?

gautierdelorme replied:
There is, but that's the problem. The [v0.19.9 release](https://github.com/go-openapi/validate/commits/v0.19.9) that includes my change also includes https://github.com/go-openapi/validate/commit/027c864a983d4f72e957c73b9236e60d7627dd30 which I don't want since it brings a lot of unneeded dep upgrades.

dims replied:
@gautierdelorme we do not encourage using SHA when there is a release tag after the SHA. in this case please use one of the release tags after the SHA. if v0.19.9 is not acceptable (say we don’t like the yaml.v2) then we need to work with the upstream project and get it fixed. we can’t pin on this SHA forever right?

dims replied:
cc @liggitt @cblecker

dims commented:
/area code-organization

thockin commented on go.mod:
```diff
@@ -172,11 +172,9 @@ replace (
 	github.com/NYTimes/gziphandler => github.com/NYTimes/gziphandler v0.0.0-20170623195520-56545f4a5d46
 	github.com/PuerkitoBio/purell => github.com/PuerkitoBio/purell v1.1.1
 	github.com/PuerkitoBio/urlesc => github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578
-	github.com/agnivade/levenshtein => github.com/agnivade/levenshtein v1.0.1
```
curious that we have (transitive) deps that are not being vendored.

@liggitt is that expected?

liggitt replied:
yeah, go.sum includes all transitive deps at the module level, go.mod pins minimum preferred version of those deps, vendor only includes the transitive packages needed to build/test

fedebongio commented:
/triage accepted

gautierdelorme commented:
As discussed in https://kubernetes.slack.com/archives/C0EG7JC6T/p1592287631273100 we moved from `go-openapi` to `kube-openapi` and we fixed the bug there (see https://github.com/kubernetes/kube-openapi/pull/213) so I'm closing this PR since the fix will be picked up by https://github.com/kubernetes/kubernetes/pull/95954.
